libraries/octree/src/OctreePacketData.cpp

```diff
@@ -743,7 +744,12 @@ int OctreePacketData::unpackDataFromBytes(const unsigned char *dataBytes, QVecto
     memcpy(&length, dataBytes, sizeof(uint16_t));
     dataBytes += sizeof(length);
     result.resize(length);
-    memcpy(result.data(), dataBytes, length * sizeof(glm::vec3));
+
+    for(int i=0;i<length;i++) {
+        memcpy(glm::value_ptr(result[i]), dataBytes, sizeof(glm::vec3));
+        dataBytes += sizeof(glm::vec3);
+    }
```

Comment:
what was wrong with all of these places that did it all at once?  it’ll be faster that way

Reply:
Recent versions of GCC emit a lot of warnings like this (this also got added to LLVM):

```
/home/dale/git/vircadia/vircadia-master/libraries/workload/src/workload/Space.cpp: In member function ‘uint32_t workload::Space::copyProxyValues(workload::Proxy*, uint32_t) const’:
/home/dale/git/vircadia/vircadia-master/libraries/workload/src/workload/Space.cpp:126:63: warning: ‘void* memcpy(void*, const void*, size_t)’ writing to an object of type ‘class workload::Proxy’ with no trivial copy-assignment; use copy-assignment or copy-initialization instead [-Wclass-memaccess]
  126 |     memcpy(proxies, _proxies.data(), numCopied * sizeof(Proxy));
      |                                                               ^
In file included from /home/dale/git/vircadia/vircadia-master/libraries/workload/src/workload/Transaction.h:21,
                 from /home/dale/git/vircadia/vircadia-master/libraries/workload/src/workload/Space.h:22,
                 from /home/dale/git/vircadia/vircadia-master/libraries/workload/src/workload/Space.cpp:15:
/home/dale/git/vircadia/vircadia-master/libraries/workload/src/workload/Proxy.h:42:7: note: ‘class workload::Proxy’ declared here
   42 | class Proxy {
      |       ^~~~~
```

The issue is that you're only supposed to memcpy over an object that's [trivially copyable](https://en.cppreference.com/w/cpp/types/is_trivially_copyable). If it isn't, this is technically undefined behaviour.

As I understand it there are two main reasons for this:

1. If there's a copy constructor, this would bypass it. That would be obviously bad.
2. There might be something other than the expected data in the class that could get overwritten, such as the vtable.

Admittedly, this probably borders on paranoia. But on the other hand, compiler authors seem to find various creative ways to optimize code that can make invoking undefined behaviour result in very weird things happening, so it might not be that bad idea to steer well clear of it.

Here's some interesting reading on how compilers react to UB: https://c9x.me/compile/bib/ubc.pdf

Reply:
interesting!  at least for structs, I would expect it to be fine and faster to do all at once

Reply:
Yes, I'm quite sure it would be faster. But, considering the compiler is telling me "writing over this with memcpy is undefined behavior", writing over an array seems like a bad idea. 

Here I added the use of glm::value_ptr seems to exist exactly for this purpose, pointing at where inside the glm::vec3 the data is found. This would seem to indicate that &vec3 may not be the same address as glm::value_ptr(vec3).

I of course welcome any better ideas here.

I'm also hoping that the compiler will optimize it back into what it was. The standard may say this is a bad idea, but the compiler knows what it generates and could well figure out it's actually safe in this particular case (I've not tried disassembling it and checking though).